### ISSUE TITLE
Deadlock in ResourceReaper when running multiple containers

### ISSUE THREAD

**Issue by mj-rittermann** [testcontainers/testcontainers-dotnet#458]
**Describe the bug**
When multiple containers are starting simultaneously, deadlock happens as the lock aquired in line 93 of ResourceReaper.cs are not released if defaultInstance is already initialized

**To Fix**
In ResourceReaper.cs move line 96-101 into the try-finally block

**Comment by mj-rittermann** [comment]
When changing to the implementation below it works:
```
public static async Task<ResourceReaper> GetAndStartDefaultAsync(CancellationToken ct = default)
    {
      if (defaultInstance != null && !defaultInstance.disposed)
      {
        return defaultInstance;
      }

      await DefaultLock.WaitAsync(ct)
        .ConfigureAwait(false);

      try
      {
        if (defaultInstance != null && !defaultInstance.disposed)
        {
          return defaultInstance;
        }

        defaultInstance = await GetAndStartNewAsync(DefaultSessionId, RyukImage, default, ct)
          .ConfigureAwait(false);

        return defaultInstance;
      }
      finally
      {
        DefaultLock.Release();
      }
    }
```

**Comment by HofmeisterAn** [comment]
👍 good catch. Thanks. I'll try to add a fix later that evening.

**Comment by HofmeisterAn** [comment]
Fixed in `1.6.0-beta.2226`.

**Comment by mj-rittermann** [comment]
Thnx - tested and working :-)
